### ISSUE TITLE
connection crash while running explain (verbose, format json) another case occured

### ISSUE THREAD

**Issue by chenbaggio** [greenplum-db/gpdb#9404]
### Greenplum version or build

v6x_stable

### OS version and uname -a

centos
Linux computing-8c091891-master-6aadb510 3.10.0-957.21.3.el7.x86_64 #1 SMP Tue Jun 18 16:35:19 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

### Expected behavior

show query plan in json format

### Actual behavior
crashed and stacks like following:

```
#0  0x00007f907ad4149b in raise () from /lib64/libpthread.so.0
#1  0x0000000000b05e56 in StandardHandlerForSigillSigsegvSigbus_OnMainThread (processName=<optimized out>, postgres_signal_arg=11) at elog.c:5586
#2  <signal handler called>
#3  show_dispatch_info (slice=0x3452830, es=es@entry=0x7ffd3489d0c0, plan=0x2603a78, plan=0x2603a78) at explain.c:971
#4  0x00000000006b63b8 in ExplainNode (planstate=0x362a448, ancestors=ancestors@entry=0x2406e00, relationship=relationship@entry=0xd37a00 "Outer", plan_name=plan_name@entry=0x0, 
    es=es@entry=0x7ffd3489d0c0) at explain.c:1624
#5  0x00000000006b64fe in ExplainNode (planstate=<optimized out>, ancestors=ancestors@entry=0x2406e00, relationship=relationship@entry=0xd37a00 "Outer", plan_name=plan_name@entry=0x0, 
    es=es@entry=0x7ffd3489d0c0) at explain.c:2278
#6  0x00000000006b64fe in ExplainNode (planstate=<optimized out>, ancestors=ancestors@entry=0x2406e00, relationship=relationship@entry=0xd37a00 "Outer", plan_name=plan_name@entry=0x0, 
    es=es@entry=0x7ffd3489d0c0) at explain.c:2278
#7  0x00000000006b64fe in ExplainNode (planstate=<optimized out>, ancestors=0x2406e00, ancestors@entry=0x0, relationship=relationship@entry=0x0, plan_name=plan_name@entry=0x0, es=es@entry=0x7ffd3489d0c0)
    at explain.c:2278
#8  0x00000000006b74f5 in ExplainPrintPlan (es=es@entry=0x7ffd3489d0c0, queryDesc=queryDesc@entry=0x2603ea0) at explain.c:775
#9  0x00000000006b7928 in ExplainOnePlan (plannedstmt=plannedstmt@entry=0x2603ca0, into=into@entry=0x0, es=es@entry=0x7ffd3489d0c0, 
    queryString=queryString@entry=0x2260d08 "EXPLAIN (VERBOSE, FORMAT JSON)\nSELECT f1 FROM INTERVAL_TBL_OF r1 ORDER BY f1;", params=params@entry=0x0, planduration=planduration@entry=0x7ffd3489d0b0)
    at explain.c:648
#10 0x00000000006b89a2 in ExplainOneQuery (into=0x0, params=0x0, queryString=0x2260d08 "EXPLAIN (VERBOSE, FORMAT JSON)\nSELECT f1 FROM INTERVAL_TBL_OF r1 ORDER BY f1;", es=0x7ffd3489d0c0, 
    query=0x36316b8) at explain.c:447
#11 ExplainQuery (stmt=stmt@entry=0x2261e00, queryString=queryString@entry=0x2260d08 "EXPLAIN (VERBOSE, FORMAT JSON)\nSELECT f1 FROM INTERVAL_TBL_OF r1 ORDER BY f1;", params=params@entry=0x0, 
    dest=dest@entry=0x3432170) at explain.c:280
#12 0x00000000009a6a39 in standard_ProcessUtility (parsetree=0x2261e00, queryString=0x2260d08 "EXPLAIN (VERBOSE, FORMAT JSON)\nSELECT f1 FROM INTERVAL_TBL_OF r1 ORDER BY f1;", 
    context=PROCESS_UTILITY_TOPLEVEL, params=0x0, dest=0x3432170, completionTag=0x7ffd3489d370 "") at utility.c:805
#13 0x00007f906c5a4cf5 in hdss_ProcessUtility (parsetree=0x2261e00, queryString=0x2260d08 "EXPLAIN (VERBOSE, FORMAT JSON)\nSELECT f1 FROM INTERVAL_TBL_OF r1 ORDER BY f1;", 
    context=PROCESS_UTILITY_TOPLEVEL, params=0x0, dest=0x3432170, completionTag=0x7ffd3489d370 "") at hashdata_stat_statements.c:775
#14 0x00000000009a3665 in PortalRunUtility (portal=portal@entry=0x2172268, utilityStmt=0x2261e00, isTopLevel=<optimized out>, dest=dest@entry=0x3432170, 
    completionTag=completionTag@entry=0x7ffd3489d370 "") at pquery.c:1381
#15 0x00000000009a427d in FillPortalStore (portal=portal@entry=0x2172268, isTopLevel=isTopLevel@entry=1 '\001') at pquery.c:1251
#16 0x00000000009a561b in PortalRun (portal=portal@entry=0x2172268, count=count@entry=9223372036854775807, isTopLevel=isTopLevel@entry=1 '\001', dest=dest@entry=0x22621b0, 
    altdest=altdest@entry=0x22621b0, completionTag=completionTag@entry=0x7ffd3489d5a0 "") at pquery.c:987
#17 0x00000000009a0314 in exec_simple_query (query_string=query_string@entry=0x2260d08 "EXPLAIN (VERBOSE, FORMAT JSON)\nSELECT f1 FROM INTERVAL_TBL_OF r1 ORDER BY f1;", 
    serializedLogicPartCtx=serializedLogicPartCtx@entry=0x0, serializedLogicPartCtxLen=serializedLogicPartCtxLen@entry=0) at postgres.c:1829
#18 0x00000000009a3033 in PostgresMain (argc=<optimized out>, argv=argv@entry=0x21970a0, dbname=0x2196f50 "postgres", username=<optimized out>) at postgres.c:5268
#19 0x0000000000544952 in BackendRun (port=0x2171d70) at postmaster.c:4930
#20 BackendStartup (port=0x2171d70) at postmaster.c:4587
#21 ServerLoop () at postmaster.c:2020
#22 0x00000000008806c5 in PostmasterMain (argc=argc@entry=6, argv=argv@entry=0x2162bf0) at postmaster.c:1560
#23 0x0000000000546f62 in main (argc=6, argv=0x2162bf0) at main.c:252

```
### Step to reproduce the behavior

1. run all the following SQLs

```
--
-- INTERVAL
--

SET DATESTYLE = 'ISO';
SET IntervalStyle to postgres;

-- check acceptance of "time zone style"
SELECT INTERVAL '01:00' AS "One hour";
SELECT INTERVAL '+02:00' AS "Two hours";
SELECT INTERVAL '-08:00' AS "Eight hours";
SELECT INTERVAL '-1 +02:03' AS "22 hours ago...";
SELECT INTERVAL '-1 days +02:03' AS "22 hours ago...";
SELECT INTERVAL '1.5 weeks' AS "Ten days twelve hours";
SELECT INTERVAL '1.5 months' AS "One month 15 days";
SELECT INTERVAL '10 years -11 month -12 days +13:14' AS "9 years...";

CREATE TABLE INTERVAL_TBL (f1 interval);

INSERT INTO INTERVAL_TBL (f1) VALUES ('@ 1 minute');
INSERT INTO INTERVAL_TBL (f1) VALUES ('@ 5 hour');
INSERT INTO INTERVAL_TBL (f1) VALUES ('@ 10 day');
INSERT INTO INTERVAL_TBL (f1) VALUES ('@ 34 year');
INSERT INTO INTERVAL_TBL (f1) VALUES ('@ 3 months');
INSERT INTO INTERVAL_TBL (f1) VALUES ('@ 14 seconds ago');
INSERT INTO INTERVAL_TBL (f1) VALUES ('1 day 2 hours 3 minutes 4 seconds');
INSERT INTO INTERVAL_TBL (f1) VALUES ('6 years');
INSERT INTO INTERVAL_TBL (f1) VALUES ('5 months');
INSERT INTO INTERVAL_TBL (f1) VALUES ('5 months 12 hours');

-- badly formatted interval
INSERT INTO INTERVAL_TBL (f1) VALUES ('badly formatted interval');
INSERT INTO INTERVAL_TBL (f1) VALUES ('@ 30 eons ago');

-- test interval operators

SELECT '' AS ten, * FROM INTERVAL_TBL;

SELECT '' AS nine, * FROM INTERVAL_TBL
   WHERE INTERVAL_TBL.f1 <> interval '@ 10 days';

SELECT '' AS three, * FROM INTERVAL_TBL
   WHERE INTERVAL_TBL.f1 <= interval '@ 5 hours';

SELECT '' AS three, * FROM INTERVAL_TBL
   WHERE INTERVAL_TBL.f1 < interval '@ 1 day';

SELECT '' AS one, * FROM INTERVAL_TBL
   WHERE INTERVAL_TBL.f1 = interval '@ 34 years';

SELECT '' AS five, * FROM INTERVAL_TBL
   WHERE INTERVAL_TBL.f1 >= interval '@ 1 month';

SELECT '' AS nine, * FROM INTERVAL_TBL
   WHERE INTERVAL_TBL.f1 > interval '@ 3 seconds ago';

SELECT '' AS fortyfive, r1.*, r2.*
   FROM INTERVAL_TBL r1, INTERVAL_TBL r2
   WHERE r1.f1 > r2.f1
   ORDER BY r1.f1, r2.f1;

-- Test intervals that are large enough to overflow 64 bits in comparisons
CREATE TEMP TABLE INTERVAL_TBL_OF (f1 interval);
INSERT INTO INTERVAL_TBL_OF (f1) VALUES
  ('2147483647 days 2147483647 months'),
  ('2147483647 days -2147483648 months'),
  ('1 year'),
  ('-2147483648 days 2147483647 months'),
  ('-2147483648 days -2147483648 months');
-- these should fail as out-of-range
INSERT INTO INTERVAL_TBL_OF (f1) VALUES ('2147483648 days');
INSERT INTO INTERVAL_TBL_OF (f1) VALUES ('-2147483649 days');
INSERT INTO INTERVAL_TBL_OF (f1) VALUES ('2147483647 years');
INSERT INTO INTERVAL_TBL_OF (f1) VALUES ('-2147483648 years');

SELECT r1.*, r2.*
   FROM INTERVAL_TBL_OF r1, INTERVAL_TBL_OF r2
   WHERE r1.f1 > r2.f1
   ORDER BY r1.f1, r2.f1;

CREATE INDEX ON INTERVAL_TBL_OF USING btree (f1);
SET enable_seqscan TO false;
EXPLAIN (COSTS OFF)
SELECT f1 FROM INTERVAL_TBL_OF r1 ORDER BY f1;



```

2. turn off optimizer

```
set optimizer=off;
```

3. run the following sql

```
EXPLAIN (VERBOSE, FORMAT JSON)
SELECT f1 FROM INTERVAL_TBL_OF r1 ORDER BY f1;

```

**Comment by Eulerizeit** [comment]
@hlinnaka is this different from the issue you fixed in https://github.com/greenplum-db/gpdb/pull/9369?

**Comment by hlinnaka** [comment]
> @hlinnaka is this different from the issue you fixed in #9369?

Seems like the same bug.

**Comment by chenbaggio** [comment]
the bug is occured after merge the code for #9369,  it seems not correct it completely

**Comment by chenbaggio** [comment]
@hlinnaka the bug is occured after merge the code for #9369, it seems not correct it completely

**Comment by hlinnaka** [comment]
> @hlinnaka the bug is occured after merge the code for #9369, it seems not correct it completely

Oh, okay.

I did some debugging, it was indeed a similar but different bug. Opened PR https://github.com/greenplum-db/gpdb/pull/9409 to fix.

**Comment by hlinnaka** [comment]
Fixed in commit 3bdece19dad997f24dd2ba7e40f11f46de3aeb24. Thanks for the report!
